components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -486,88 +486,363 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
+     * @param limit              Number of required results (count)
+     * @param sortBy             Sortby
+     * @param sortOrder          Sort order
+     * @param domainName         Domain that the filter should perform
+     * @return Detailed user list
      * @throws CharonException
      */
     private List<Object> filterUsers(Node node, Map<String, Boolean> requiredAttributes, int offset, int limit,
-                                     String sortBy, String sortOrder, String domainName)
-            throws NotImplementedException, CharonException {
+            String sortBy, String sortOrder, String domainName) throws CharonException {
 
-        List<Object> filteredUsers = new ArrayList<>();
-        //0th index is to store total number of results
-        filteredUsers.add(0);
-        String[] userNames;
+        if (!(limit > 0))
+            limit = getMaxLimit();
 
         //Handle single attribute search
         if (node instanceof ExpressionNode) {
-            String attributeName = ((ExpressionNode) node).getAttributeValue();
-            String filterOperation = ((ExpressionNode) node).getOperation();
-            String attributeValue = ((ExpressionNode) node).getValue();
+            return getSingleAttributeFilteredUsers((ExpressionNode) node, requiredAttributes, offset, limit, sortBy,
+                    sortOrder, domainName);
 
-            if (log.isDebugEnabled()) {
-                log.debug(String.format("Listing users by filter: %s %s %s",
-                        attributeName, filterOperation, attributeValue));
-            }
+        } else if (node instanceof OperationNode) {
 
-            try {
-                if (isNotFilteringSupported(filterOperation)) {
-                    throw new NotImplementedException(String.format("System does not support filter operator: %s",
-                            filterOperation));
-                }
+            List<Object> filteredUsers = new ArrayList<>();
+            // 0th index is to store total number of results.
+            filteredUsers.add(0);
 
-                if (SCIMCommonUtils.isFilteringEnhancementsEnabled()) {
-                    if (SCIMCommonConstants.EQ.equalsIgnoreCase(filterOperation)) {
-                        if (StringUtils.equals(attributeName, SCIMConstants.UserSchemaConstants.USER_NAME_URI) &&
-                                !StringUtils.contains(attributeValue, CarbonConstants.DOMAIN_SEPARATOR)) {
-                            attributeValue = UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME +
-                                    CarbonConstants.DOMAIN_SEPARATOR + attributeValue;
-                        }
-                    }
+            // Support multi attribute filtering.
+            return getMultiAttributeFilteredUsers(node, requiredAttributes, offset, limit, sortBy, sortOrder,
+                    domainName, filteredUsers);
+        } else {
+            throw new CharonException("Unsupported Operation");
```

Comment:
If we have any context information on the operation let's add them here

Reply:
fixed : fda144b175b11a4dabab023d9bd8b3f897a7a16c